### PR TITLE
avoid tf.int32-only call to gather in moments, thereby keeping kernel on GPU

### LINKED ISSUE #6027: tracing batch normalization: apparently unnecessary gpu-cpu-gpu transfer for gather op.
I am running:
CUDA 8.0. CUDNN 5.1. tensorflow 0.11. Ubuntu 12.04.

I was experiencing low GPU utilization. As per #1824 I did a trace and found many instances of the following interaction, as shown by the screenshot in chrome:

The op I have highlighted (with arrows going in and out in the screenshot) is a `gather` op. I provide a link to the context in which it is called.
https://github.com/tensorflow/tensorflow/blob/eb56a8af24695bf8258addf28b0c53fbabff72e6/tensorflow/python/ops/nn_impl.py#L472

The call stack (at graph construction) time is:
`tf.contrib.layers.batch_norm`, `python.ops.nn_impl.moments`, `python.ops.nn_impl.sufficient_statics`

If I understand correctly, I am seeing that the GPU gets blocked as a result of shuffling a trivial `gather` op unnecessarily off to the CPU only to wait forever for it to come back. 

screenshot from 2016-12-01 11 45 17

### PR DESCRIPTION
**Pull request by liuyipei**
Fixes #6027 (tested locally)
@vincentvanhoucke 
Sorry for being late here. I think this might be a slightly better fix. I tested this locally, and it did resolve the cpu-gather behavior that I was seeing.
@poxvoculi 
I thought about it for a while, and I couldn't figure out how to avoid `tf.gather` completely in a reasonable manner. I hope that this patch is acceptable.

### CODE REVIEW

**Comment by vincentvanhoucke** [comment]
Prefer #6031 to this equivalent fix. Reviewers please comment on that PR. @liuyipei thanks for raising the issue!
